README.md

```diff
@@ -1,6 +1,93 @@
-## History
+**Look: The Generic Distributed Materialized Views Engine.**
+
+_By engineers, for engineers, crowdsourced under the umbrella of the SysDesign Meetup._
+
+## Product Vision
+
+Look is the ultimate high-performance materialized view.
+
+It aggregates data mutations from various sources, applies these mutations to a single large, holistic, mostly-in-memory data blob, and exposes interfaces to run queries against this data, at its most recent snapshot as well as against its past.
+
+Look is designed to handle millions of mutations and read requests per second on a single node. At the same time, Look is distributed-first, durable, and infinitely scalable, and its design emphasises zero downtime in addition to high throughput.
+
+The only requirement for the data sources is to provide their updates in a stream. This stream of updates should have the property of retaining each update message for a short period of time (single-digit seconds, minutes max), so that each update is guaranteed to be delivered to Look. Three most common use cases that immediately fit the bill are the Transactional Outbox RDBMS pattern, Kafka streams, or a Git branch with force-pushes disabled.
+
+When multiple input data sources need to be joined, the user does not have to worry about the relative order of events in them. Specifically, this applies to feeding Look from a Kafka topic with multiple partitions. Look will never violate the order of updates that have been pushed to it via each individual channel, and the data in Look will always remain as consistent as it was at the point of entry. Look can further be used to increase the degree of consistency of the data served by it, by allowing the users to implement custom logic for the respective CQRS queries, so that this logic can validate user-defined invariants before sending the response to the query.
+
+## Tech Philosophy
+
+TL;DR specific to Look:
+
+* Easy to run, test, and use on a single machine, both in dev and in prod setting.
+* Interfacing to the constellation of nodes is identical to interfacing with a single node.
+* Infinitely scalable means can add nodes on the fly, with no downtime.
+
+For a more complete description, please refer to [this section](https://github.com/SysDesignMeetup/sdm#educational-designs) of SysDesignMeetup's top-level README on our crowdsourced designs.
+
+## Implementation Challenges
+
+There are four pillars to Look.
+
+### Linearizing Vector Clocks
+
+In terms of functional programming, the inputs to Look are monadic transormations. Transactional Outboxes, Kafka consumers, logical replication logs, CDC pipelines and Git commits can all be viewed as such.
+
+The inputs do not have to be synchronized with one another; even Kafka partitions within the same topic are not synchronized. Thus, the spectrum of inputs to a logical instance of Look can be described by a vector clock. Inside Look, data updates are applied in a strictly sequential, totally ordered, linearized way.
+
```

Comment:
Do we plan to provide some configured priority of various inputs? Or the order can be more or less arbitrary?

Reply:
I thought of this, and my current thinking is that it would be too much.

In theory, if the source of data is a Kafka topic of multiple partitions, it is relatively easy to implement the logic that each partition should contain heartbeat messages, say, every second. Then, an invariant could be enforced that the state of the data we have has at most two or three seconds of "drift" between these partitions.

In practice, it's generally a source of hard-to-find bugs. It's enough to make this high discrepancy a paging alert, IMHO.